## Task 1.1: w/o Pandas and Numpy

In [35]:
# Function to read data from CSV
def read_csv(path):
    try:
        with open(path, "r") as file:
            rows = []
            for line in file:
                row = line.strip().split(",")
                rows.append(row)
            return rows
    except Exception as e:
        print(f"An error occured while reading {path}: {e}")

In [47]:
# Read data from disk using read_csv function
train_in = read_csv("./resources/train_in.csv")
train_out = read_csv("./resources/train_out.csv")

In [54]:
# Transform train_in
transformed_train_in = [[float(num) for num in sublist] for sublist in train_in]

In [96]:
# Transform train_out
transformed_train_out = [int(num[0]) for num in train_out]

In [62]:
# Aggregate the 0's, 1's etc
centres_aggregate = { 0: [], 1: [], 2: [], 3: [], 4: [], 5: [], 6: [], 7: [], 8: [], 9: [] }

for i in range(len(train_in)):
    centres_aggregate[transformed_train_out[i]].append(transformed_train_in[i])

In [65]:
# Sum the 0's, 1's etc
centres_total = { 0: None, 1: None, 2: None, 3: None, 4: None, 5: None, 6: None, 7: None, 8: None, 9: None }

for i in centres_aggregate:
    total = [sum(values) for values in zip(*centres_aggregate[i])]
    centres_total[i] = total

In [74]:
# Average of 0's, 1's etc
centres_average = { 0: None, 1: None, 2: None, 3: None, 4: None, 5: None, 6: None, 7: None, 8: None, 9: None }

for i in centres_aggregate:
    average = [values/len(centres_aggregate[i]) for values in centres_total[i]]
    centres_average[i] = average

In [103]:
# Mean the 0's, 1's etc
centres_mean = { 0: None, 1: None, 2: None, 3: None, 4: None, 5: None, 6: None, 7: None, 8: None, 9: None }

for i in centres_aggregate:
    mean = [sum(values)/len(centres_aggregate[i]) for values in zip(*centres_aggregate[i])]
    centres_mean[i] = mean

In [124]:
import math

# Generate the distance matrix (Euclidean distance)
distance_matrix = []
for i in centres_mean:
    row_i = []
    for j in centres_mean:
        difference_i_j = [abs(a - b) for a, b in zip(centres_mean[i], centres_mean[j])]
        square_i_j = [x * x for x in difference_i_j]
        distance_i_j = math.sqrt(sum(square_i_j))
        row_i.append(distance_i_j)
    distance_matrix.append(row_i)

In [130]:
import math

# Simplified
# Generate the distance matrix (Euclidean distance)
distance_matrix = []
for i in centres_mean:
    row_i = []
    for j in centres_mean:
        distance_i_j = math.sqrt(sum((a - b) ** 2 for a, b in zip(centres_mean[i], centres_mean[j])))
        row_i.append(distance_i_j)
    distance_matrix.append(row_i)

In [141]:
# Print the matrix
print("        ".join(str(i) for i in range(10)))
for index, row in enumerate(distance_matrix):
    formatted_row = [f"{element:.2f}" for element in row]
    print(f"{index}  ")
    print("    ".join(formatted_row))

0        1        2        3        4        5        6        7        8        9
0  
0.00    14.45    9.33    9.14    10.77    7.52    8.15    11.86    9.91    11.49
1  
14.45    0.00    10.13    11.73    10.17    11.12    10.61    10.74    10.09    9.93
2  
9.33    10.13    0.00    8.18    7.93    7.91    7.33    8.87    7.08    8.89
3  
9.14    11.73    8.18    0.00    9.09    6.12    9.30    8.92    7.02    8.35
4  
10.77    10.17    7.93    9.09    0.00    8.00    8.78    7.58    7.38    6.01
5  
7.52    11.12    7.91    6.12    8.00    0.00    6.70    9.21    6.97    8.26
6  
8.15    10.61    7.33    9.30    8.78    6.70    0.00    10.89    8.59    10.44
7  
11.86    10.74    8.87    8.92    7.58    9.21    10.89    0.00    8.47    5.43
8  
9.91    10.09    7.08    7.02    7.38    6.97    8.59    8.47    0.00    6.40
9  
11.49    9.93    8.89    8.35    6.01    8.26    10.44    5.43    6.40    0.00


- 5 and 3 look similar, 6.12
- 8 and 9 look similar, 6.70

## Task 1.1: w Pandas and Numpy and Scipy

In [206]:
# Install pandas and numpy
!pip install pandas numpy scipy

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.8/60.8 kB 2.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.1/23.1 MB 24.8 MB/s eta 0:00:00m eta 0:00:010:01:01

[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: pip install --upgrade pip


In [149]:
# Import dependencies
import pandas as pd
import numpy as np

In [150]:
# Read data from disk using pandas
train_in = pd.read_csv("./resources/train_in.csv")
train_out = pd.read_csv("./resources/train_out.csv")

In [192]:
# Combining both DataFrames
combined_data = pd.concat([train_in, train_out], axis = 1)

In [197]:
# Finding the centre (average after grouping) and removing the last column
centre_data = combined_data.groupby(combined_data.iloc[:, 256]).mean().iloc[:, :-1]

In [213]:
# Find the Euclidean distance between two vectors
distance = np.linalg.norm(centre_data.iloc[0] - centre_data.iloc[0])

In [208]:
# Create distance matrix using scipy
from scipy.spatial import distance_matrix

dij = distance_matrix(centre_data.values, centre_data.values)

[[ 0.         14.44960797  9.33455587  9.14373367 10.76984444  7.51929626
   8.16301949 11.86455505  9.90790174 11.48887494]
 [14.44960797  0.         10.12532258 11.7332329  10.17378643 11.11880041
  10.59793273 10.74315367 10.08677677  9.9320937 ]
 [ 9.33455587 10.12532258  0.          8.17828489  7.93254148  7.90679632
   7.31795234  8.87253107  7.07751618  8.88774785]
 [ 9.14373367 11.7332329   8.17828489  0.          9.0876078   6.11875002
   9.29242591  8.92240093  7.02042489  8.35435012]
 [10.76984444 10.17378643  7.93254148  9.0876078   0.          8.00151741
   8.76599667  7.58301228  7.38090899  6.01040793]
 [ 7.51929626 11.11880041  7.90679632  6.11875002  8.00151741  0.
   6.68862915  9.21195402  6.96738648  8.25853807]
 [ 8.16301949 10.59793273  7.31795234  9.29242591  8.76599667  6.68862915
   0.         10.8681573   8.57020763 10.41974352]
 [11.86455505 10.74315367  8.87253107  8.92240093  7.58301228  9.21195402
  10.8681573   0.          8.4677853   5.42647412]
 [ 9.907

In [212]:
# Convert the matrix back into a DataFrame
distance_df = pd.DataFrame(dij, index=[f"R: {i}" for i in range(10)], columns=[f"C: {i}" for i in range(10)])

print(distance_df)

           C: 0       C: 1       C: 2       C: 3       C: 4       C: 5  \
R: 0   0.000000  14.449608   9.334556   9.143734  10.769844   7.519296   
R: 1  14.449608   0.000000  10.125323  11.733233  10.173786  11.118800   
R: 2   9.334556  10.125323   0.000000   8.178285   7.932541   7.906796   
R: 3   9.143734  11.733233   8.178285   0.000000   9.087608   6.118750   
R: 4  10.769844  10.173786   7.932541   9.087608   0.000000   8.001517   
R: 5   7.519296  11.118800   7.906796   6.118750   8.001517   0.000000   
R: 6   8.163019  10.597933   7.317952   9.292426   8.765997   6.688629   
R: 7  11.864555  10.743154   8.872531   8.922401   7.583012   9.211954   
R: 8   9.907902  10.086777   7.077516   7.020425   7.380909   6.967386   
R: 9  11.488875   9.932094   8.887748   8.354350   6.010408   8.258538   

           C: 6       C: 7       C: 8       C: 9  
R: 0   8.163019  11.864555   9.907902  11.488875  
R: 1  10.597933  10.743154  10.086777   9.932094  
R: 2   7.317952   8.872531   7.0